src/targets/http.rs

```diff
@@ -76,33 +96,106 @@ impl Target {
                     "Target {}: Got update ({} entries)",
                     component.name(), update.set().len()
                 );
-                source.update(update);
+                source.update(SourceData::new(update, &mut state));
             }
         }
     }
 }
 
 
-
 //------------ Source --------------------------------------------------------
 
 /// The date source for an HTTP target.
 #[derive(Clone, Default)]
 struct Source {
     /// The current set of RTR data.
-    data: Arc<ArcSwap<Option<payload::Set>>>
+    data: Arc<ArcSwap<Option<SourceData>>>
 }
 
 impl Source {
     /// Updates the data source from the given update.
-    fn update(&self, update: payload::Update) {
-        self.data.store(Some(update.set().clone()).into())
+    fn update(&self, data: SourceData) {
+        self.data.store(Some(data).into())
     }
 
-    /// Returns the current payload set.
-    fn set(&self) -> Option<payload::Set> {
-        self.data.load().as_ref().as_ref().cloned()
+    /// Returns the current payload data.
+    fn data(&self) -> Arc<Option<SourceData>> {
+        self.data.load_full()
     }
 }
 
 
+//------------ SourceData ----------------------------------------------------
+
+/// The data held by a data source.
+///
+struct SourceData {
+    set: payload::Set,
+    etag: String,
+    created: DateTime<Utc>,
+}
+
+impl SourceData {
+    fn new(update: payload::Update, state: &mut State) -> Self {
+        let etag = format!("\"{:x}-{}\"", state.session(), state.serial());
+        state.inc();
+        Self {
+            set: update.set().clone(),
+            etag,
+            created: Utc::now(),
+        }
+    }
+
+    /// Returns whether 304 Not Modified response should be retured.
+    fn is_not_modified(&self, req: &Request<Body>) -> bool {
+        // First, check If-None-Match.
+        for value in req.headers().get_all("If-None-Match").iter() {
```

Comment:
It's unclear to me from a glance at the code if the get("Some-Header-Name") calls in this fn are case-insensitive or not. https://datatracker.ietf.org/doc/html/rfc2616#section-4.2 says that header names are case-insensitive so we shouldn't require a specific case match here.

Reply:
Hyper (or possibly the underlying _http_ crate) convert standard names into a variant of an enum. What we should do is use the constants from `hyper::header` instead of a string. I’ll fix that.